### PR TITLE
WIP: [Issue 218] Add Seek functions to Reader

### LINKED ISSUE #218: Seek function is missing in Reader
`Seek()` and `SeekByTime()` are available in `Consumer` but not `Reader`. The C++ Client has the function in both classes.

### PR DESCRIPTION
**Pull request by cornelk**
Fixes #218 

WIP due to `TestReaderSeek` hanging when executing `reader.Next()` after `reader.Seek()` - not sure why this happens and how to continue here, any help or hint is welcome.

### Motivation

`Seek()` was missing in `Reader` interface but is available in the C++ client lib.

### Modifications

* implemented Seek in the similar way as `Consumer` does

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

- `TestReaderSeek*`

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API: (**yes**)
  - The schema: (don't know)
  - The default values of configurations: (no)
  - The wire protocol: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (GoDocs)

### CODE REVIEW

**Comment by wolfstudy** [comment]
```
panic: test timed out after 20m0s
```

**Comment by wolfstudy** [comment]
Move this change to 0.2.0

File: pulsar/reader_impl.go [hunk 1]
```diff
@@ -150,3 +151,33 @@ func (r *reader) hasMoreMessages() bool {
...
+
+	partition := mid.partitionIdx
+	// did we receive a valid partition index?
+	if partition != 0 {
+		r.log.Warnf("invalid partition index %d expected 0", partition)
+		return nil, false
+	}
+
+	return mid, true
+}
```
Comment (nitishv) [inline pulsar/reader_impl.go:170]: Given the method name, the purpose of this method is to convert MessageID to get *messageID
In doing so failure in type assertion is not really a failure. The given msgID can still be converted. Refer to a this fix I just made https://github.com/apache/pulsar-client-go/pull/305

I think that change can be re-factored into this unexported method for reader, as it being used in couple of places now with this change.
Also checking for non-zero partition index does not seem to be the purpose of this method, and should not be included here. That check is only relevant to Seek APIs, and should be moved there.

File: pulsar/reader_test.go [hunk 2]
```diff
@@ -362,3 +365,108 @@ func TestReaderHasNext(t *testing.T) {
 
 	assert.Equal(t, 10, i)
 }
+
+func TestReaderSeek(t *testing.T) {
```
Comment (nitishv) [inline pulsar/reader_test.go:369]: There can be an added test case which does a Seek on a custom MessageID refer to https://github.com/apache/pulsar-client-go/pull/305

**Comment by wolfstudy** [comment]
ping @cornelk Can you check the @nitishv comments and fix the logic of `seek` for reader.

**Comment by cornelk** [comment]
> ping @cornelk Can you check the @nitishv comments and fix the logic of `seek` for reader.

pong. I am focusing on my own Pulsar Go Client now, someone else can take over this MR

**Pull request closed without merging**